src/model/gateway.rs

```diff
@@ -549,107 +548,86 @@ pub struct ClientStatus {
-            .remove("user")
-            .ok_or_else(|| DeError::custom("expected presence user"))
-            .and_then(JsonMap::deserialize)
-            .map_err(DeError::custom)?;
-
-        let (user_id, user) = if user_map.len() > 1 {
-            let user = User::deserialize(Value::from(user_map)).map_err(DeError::custom)?;
-
-            (user.id, Some(user))
-        } else {
-            let user_id = user_map
-                .remove("id")
-                .ok_or_else(|| DeError::custom("Missing presence user id"))
-                .and_then(UserId::deserialize)
-                .map_err(DeError::custom)?;
-
-            (user_id, None)
-        };
-
-        let activities = match map.remove("activities") {
-            Some(v) => from_value::<Vec<Activity>>(v).map_err(DeError::custom)?,
-            None => Vec::new(),
-        };
-
-        let client_status = match map.remove("client_status") {
-            Some(v) => from_value::<Option<ClientStatus>>(v).map_err(DeError::custom)?,
-            None => None,
-        };
-
-        let last_modified = match map.remove("last_modified") {
-            Some(v) => from_value::<Option<u64>>(v).map_err(DeError::custom)?,
-            None => None,
-        };
+impl PresenceUser {
+    /// Attempts to convert this [`PresenceUser`] instance into a [`User`].
+    ///
+    /// If one of [`User`]'s required fields is None in `self`, None is returned.
+    pub fn into_user(self) -> Option<User> {
+        Some(User {
+            avatar: self.avatar,
+            bot: self.bot?,
+            discriminator: self.discriminator?,
+            id: self.id,
+            name: self.name?,
+            public_flags: self.public_flags,
+        })
+    }
 
-        let status = map
-            .remove("status")
-            .ok_or_else(|| DeError::custom("expected presence status"))
-            .and_then(OnlineStatus::deserialize)
-            .map_err(DeError::custom)?;
-
-        Ok(Presence {
-            activities,
-            client_status,
-            last_modified,
-            status,
-            user_id,
-            user,
+    /// Attempts to convert this [`PresenceUser`] instance into a [`User`].
+    ///
+    /// Will clone individual fields if needed.
+    ///
+    /// If one of [`User`]'s required fields is None in `self`, None is returned.
+    pub fn to_user(&self) -> Option<User> {
+        Some(User {
+            avatar: self.avatar.clone(),
+            bot: self.bot?,
+            discriminator: self.discriminator?,
+            id: self.id,
+            name: self.name.clone()?,
+            public_flags: self.public_flags,
         })
     }
-}
 
-impl Serialize for Presence {
-    fn serialize<S>(&self, serializer: S) -> StdResult<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        #[derive(Serialize)]
-        struct UserId {
-            id: u64,
-        }
+    // TODO: add `to_current_user` and `into_current_user`? (converts to Option<CurrentUser>)
```

Comment:
While I don't find the significance of these (how often would they be used?), I also don't see harm in adding the methods. `PresenceUser` does structurally match `CurrentUser`, with the only difference between them being that the former has all but one optional fields. If Discord serves all fields of `PresenceUser`, it can be losslessly converted into `CurrentUser`.

Reply:
If you don't see the significance of the methods, I'll just not add them. Less complexity that way :)